Lambda instrumentation implementation

bonybrown opened the pull request:
*Issue #20*

*Description of changes:* 
* Adds `aws-xray-sdk/lambda.rb` to be used in lambda functions to initialise and configure XRay.recorder.
* Adds subclasses of existing Default Context, Emitter and Streamer classes to customise these object's behaviours in the lambda context
* Adds a facade Segment object that the lambda uses as the root segment for all subsegments recorded in the lambda.

This work is based off the implementation of the python xray-sdk for lambda.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. ✔️

srprash commented on lib/aws-xray-sdk/lambda/facade_segment.rb:
```diff
@@ -0,0 +1,13 @@
+module XRay
+  class FacadeSegment < XRay::Segment
```
FacadeSegment should restrict any mutation to itself except adding/removing subsegments and should raise `UnsupportedOperation` exception. Can you add those methods too? Please refer to the [python SDK FacadeSegment](https://github.com/aws/aws-xray-sdk-python/blob/master/aws_xray_sdk/core/models/facade_segment.py) for details.

bonybrown replied:
Done

srprash commented on lib/aws-xray-sdk/lambda/lambda_emitter.rb:
```diff
@@ -0,0 +1,15 @@
+  # when the function execution is complete.
+  class LambdaEmitter < XRay::DefaultEmitter
+    def should_send?(entity:)
+      entity.name != '127.0.0.1' #Do not send localhost entities. It's the ruby lambda runtime
```
Can you explain what entity is this and why shouldn't it be streamed?

bonybrown replied:
When the `net_http` library is instrumented, segments appear in the traces that seem to be the ruby lambda runtime sending back the results of the lambda invocation to the lambda service.

They appear in the x-ray traces like this:
![image](https://user-images.githubusercontent.com/1063621/63632988-67256f00-c684-11e9-8007-03b58e5bc164.png)

When segments from `127.0.0.1` are not sent, the trace looks like:
![image](https://user-images.githubusercontent.com/1063621/63633035-d26f4100-c684-11e9-8638-309bd6de9404.png)

Which seems to be more expected ( ie, user would not expect to see the lambda runtime post in the x-ray trace). The alternative is to modify the instrumentation of the `net_http` library so that it (somehow) does not instrument the AWS lambda runtime calls?

Here's the section of the raw trace that the post appears in:
```
{
            "Document": {
                "id": "321193b7061fddab",
                "name": "127.0.0.1",
                "start_time": 1566623930.7746437,
                "end_time": 1566623930.775297,
                "parent_id": "f594bde1a1f24670",
                "inferred": true,
                "http": {
                    "request": {
                        "url": "http://127.0.0.1:9001/2018-06-01/runtime/invocation/0cf18128-f648-4379-b7fc-2fa4f07d5ed6/response",
                        "method": "POST"
                    },
                    "response": {
                        "status": 202,
                        "content_length": 16
                    }
                },
                "trace_id": "1-5d60c8ba-8227af6a5bfdc58b778dab40"
            },
            "Id": "321193b7061fddab"
        },
```

bonybrown replied:
Matching on the `entity.name` seems broad, but there's not much other information in the entity that would help to identify the context it is being sent from.

bonybrown replied:
Alternative is to not capture the subsegment by modifying the `net_http` patch around https://github.com/aws/aws-xray-sdk-ruby/blob/master/lib/aws-xray-sdk/facets/net_http.rb#L24 to match requests made by the lambda runtime?

Something like:
```ruby
      LAMBDA_RUNTIME_PATH_PATTERN = /\A\/[0-9-]{10}\/runtime\/invocation\/[0-9a-f-]{36}\/response\Z/.freeze
      def request(req, body = nil, &block)
        if req.path && (req.path == ('/GetSamplingRules') || req.path == ('/SamplingTargets')
                        || req.path =~ LAMBDA_RUNTIME_PATH_PATTERN)
          return super
        end
```

bonybrown replied:
My suspicion is that it is because the Ruby runtime is based on the [custom runtime](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-custom.html) work and the [runtime api](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html). 
The other runtimes that were available before Ruby may not use the runtime api?

srprash replied:
Okay. This could be the reason. I believe we can accept this and not capture the localhost subsegment by using the approach you mentioned in the comment above. Can you make that change as well?

bonybrown commented:
Also, created this project that deploys a Lambda function that demonstrates the forked gem.
https://github.com/bonybrown/aws-lambda-ruby-xray-demo

bonybrown commented on lib/aws-xray-sdk/lambda/lambda_context.rb:
```diff
@@ -0,0 +1,37 @@
+  # LambdaContext extends the default context so that
+  # we can provide an appropriate FacadeSegment as the
+  # root context for each function invocation.
+  class LambdaContext < XRay::DefaultContext
```
This is done and tests written.

bonybrown commented:
All requested changes made and confirmed to be working in https://github.com/bonybrown/aws-lambda-ruby-xray-demo.

srprash commented:
Hi @bonybrown ,
Great work. I really appreciate your contribution. I tested the code on my end and also checked out the [demo app](https://github.com/bonybrown/aws-lambda-ruby-xray-demo) you put together. 
I am experience a little issue with both. When I disable the **Active tracing** from the lambda function console, I still see an empty trace generated in the X-Ray console. Please see the following images.

![Screen Shot 2019-09-09 at 2 59 07 PM](https://user-images.githubusercontent.com/50466688/64573521-6f3a0b80-d320-11e9-9dca-b82fd73e5ac1.png)

And this is the trace data:
![Screen Shot 2019-09-09 at 2 59 30 PM](https://user-images.githubusercontent.com/50466688/64573551-8d077080-d320-11e9-8b22-aec7eaa8aee0.png)

Are you seeing this empty trace as well?

bonybrown commented:
Just tried, and yes, I'm seeing the same behaviour. Interesting that with Active Tracing disabled in the console, the REPORT output still shows an XRay trace id:
```
END RequestId: 7238dcbb-84d7-401f-943c-fff00b1c10da
REPORT RequestId: 7238dcbb-84d7-401f-943c-fff00b1c10da	Duration: 259.67 ms	Billed Duration: 300 ms	Memory Size: 128 MB	Max Memory Used: 55 MB	
XRAY TraceId: 1-5d76e4fe-4481878076dd858f6f8c4547	SegmentId:
```
I'll investigate and get back to you.

bonybrown commented:
Hi @srprash  - problem identified and fixed. In doing so, identified another problem, now also fixed.
Tested with Active Tracing turned on AND off.
No longer creating empty traces when the lambda is not sampled.

bonybrown commented:
Hi @srprash, any update on the progress of this PR? I believe all the issues raised have been addressed.

srprash commented:
Hi @bonybrown ,
Sorry for the delay. We are talking to the internal team to see if there's a fix for the localhost entity that we are seeing in the lambda function so that we can avoid the workaround in the PR. Once, it is sorted, we can get this PR merged.

Thanks,
Prashant

srprash commented on lib/aws-xray-sdk/facets/net_http.rb:
```diff
@@ -16,12 +16,18 @@ def new(*options)
 
+      # Pattern matches the path of requests from the AWS Lambda Ruby Runtime.
+      # We do not want to capture these
+      LAMBDA_RUNTIME_PATH_PATTERN = /\A\/[0-9-]{10}\/runtime\/invocation\//.freeze
```
Instead of using regex for pattern matching, I think we can fetch the value of `AWS_LAMBDA_RUNTIME_API` environment variable and check if the req.path begins with this value. I just came across this env variable in [lambda runtime](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html) and we can avoid using regex matching altogether.

srprash replied:
`req.path` would only give the partial path that does not include the host information. Instead `req.uri` gives the complete url `http://127.0.0.1:9001/.....`. Doing a `start_with?` check on `req.uri` would work better. What do you think?

bonybrown replied:
Changed to test against the `AWS_LAMBDA_RUNTIME_API` environment variable.
Tested in sample app and seems to be working fine.

srprash commented on lib/aws-xray-sdk/lambda/lambda_recorder.rb:
```diff
@@ -0,0 +1,19 @@
+  class LambdaRecorder < Recorder
+    include Logging
+
+    def begin_segment(name, trace_id: nil, parent_id: nil, sampled: nil)
```
Not sure if this works, since I'm able to create a segment using `XRay.recorder.begin_segment 'my_segment`. Can you confirm this as well?

bonybrown replied:
Added test to demo project, and seems to be working as expected ( logs warning, returns current Facade Segment)

srprash replied:
I see. Some problem on my end.

bonybrown commented:
Sorry, seems that my demo project was using the prior version of `aws-xray-sdk-ruby` - note to self: it's important to completely remove the `vendor` directory when changing versions of gems from git.

Have fixed and added tests of `begin_segment` and `begin_subsegment` to the demo project.
```ruby
 # Test begin_segment - should report warning, but not create segment
XRay.recorder.begin_segment 'my_segment'
puts "in segment - begin_segment call should have logged a warning"
# Test end_segment - should report warning
XRay.recorder.end_segment

# Test begin_subsegment, should result in subsegment created
XRay.recorder.begin_subsegment 'sub-segment'
puts "in subsegment"
dynamo_client.list_tables
uri = URI 'https://ip-ranges.amazonaws.com/ip-ranges.json'
http_connection.request(uri).body
XRay.recorder.end_subsegment
```

Using aws-xray-sdk 0.11.2 from https://github.com/bonybrown/aws-xray-sdk-ruby.git (at lambda-instrumentation@d2d7223) and updated demo project I now get this output:
```
START RequestId: 7301fb2e-249d-45e7-a468-51167a136a33 Version: $LATEST
W, [2019-10-07T23:45:57.171513 #8]  WARN -- : Cannot create segments inside Lambda function. Returning current segment.
in segment - begin_segment call should have logged a warning
W, [2019-10-07T23:45:57.186110 #8]  WARN -- : Cannot end segment inside Lambda function. Ignored.
in subsegment
END RequestId: 7301fb2e-249d-45e7-a468-51167a136a33
REPORT RequestId: 7301fb2e-249d-45e7-a468-51167a136a33	Duration: 996.39 ms	Billed Duration: 1000 ms	Memory Size: 128 MB	Max Memory Used: 64 MB	Init Duration: 927.05 ms	
XRAY TraceId: 1-5d9bce34-db783d334fd2d50cabd9e15a	SegmentId: 0bc8f498226155ce	Sampled: true	
```
I don't get `ERROR -- : can not find the current context.` at all, and attempting to begin a segment results in a warning log message (the existing facade segment is returned) as expected.

bonybrown commented on lib/aws-xray-sdk/facets/net_http.rb:
```diff
@@ -20,8 +20,21 @@ def initialize(*options)
+      def lambda_runtime_request?(req)
+        ENV['AWS_LAMBDA_RUNTIME_API']  &&
+          req.uri &&
+          req.uri.to_s.starts_with?('http://'+ENV['AWS_LAMBDA_RUNTIME_API']+'/')
```
Fixed.

bonybrown commented:
This is the x-ray trace for the invocation logged above ( TraceId: 1-5d9bce34-db783d334fd2d50cabd9e15a )
![image](https://user-images.githubusercontent.com/1063621/66357483-d0560d00-e9ba-11e9-8d91-0335af1dfd2a.png)

srprash commented:
Looks good and ready to be merged!
Thanks for your contribution.
